File: parser.py
from tokenizer import Token


class ASTNode:
    pass


class Literal(ASTNode):
    def __init__(self, value):
        self.value = value

    def __repr__(self):
        return f"Literal({self.value})"


class Identifier(ASTNode):
    def __init__(self, name):
        self.name = name

    def __repr__(self):
        return f"Identifier({self.name})"


class ArrayLiteral(ASTNode):
    def __init__(self, elements):
        self.elements = elements

    def __repr__(self):
        return f"ArrayLiteral({self.elements})"


class ArrayAccess(ASTNode):
    def __init__(self, array, index):
        self.array = array
        self.index = index

    def __repr__(self):
        return f"ArrayAccess({self.array}, {self.index})"


class BinaryOp(ASTNode):
    def __init__(self, left, operator, right):
        self.left = left
        self.operator = operator
        self.right = right

    def __repr__(self):
        return f"BinaryOp({self.left}, {self.operator}, {self.right})"


class UnaryOp(ASTNode):
    def __init__(self, operator, operand):
        self.operator = operator
        self.operand = operand

    def __repr__(self):
        return f"UnaryOp({self.operator}, {self.operand})"


class Assignment(ASTNode):
    def __init__(self, target, value):
        self.target = target
        self.value = value

    def __repr__(self):
        return f"Assignment({self.target}, {self.value})"


class IfStatement(ASTNode):
    def __init__(self, condition, then_block, else_block=None):
        self.condition = condition
        self.then_block = then_block
        self.else_block = else_block

    def __repr__(self):
        return f"IfStatement({self.condition}, {self.then_block}, {self.else_block})"


class WhileLoop(ASTNode):
    def __init__(self, condition, body):
        self.condition = condition
        self.body = body

    def __repr__(self):
        return f"WhileLoop({self.condition}, {self.body})"


class ForLoop(ASTNode):
    def __init__(self, assignment, end, body):
        self.assignment = assignment
        self.end = end
        self.body = body

    def __repr__(self):
        return f"ForLoop({self.assignment}, {self.end}, {self.body})"


class ReturnStatement(ASTNode):
    def __init__(self, value):
        self.value = value

    def __repr__(self):
        return f"ReturnStatement({self.value})"


class Block(ASTNode):
    def __init__(self, statements):
        self.statements = statements

    def __repr__(self):
        return f"Block({self.statements})"


class Parser:
    def __init__(self, tokens: list[tuple[str, Token]]):
        self.tokens = tokens
        self.pos = 0

    def current_token(self) -> tuple[str, Token] | None:
        if self.pos < len(self.tokens):
            return self.tokens[self.pos]
        return None

    def peek_token(self, offset=1) -> tuple[str, Token] | None:
        if self.pos + offset < len(self.tokens):
            return self.tokens[self.pos + offset]
        return None

    def advance(self):
        self.pos += 1

    def consume(self, expected_value=None) -> tuple[str, Token]:
        token = self.current_token()
        if token is None:
            raise SyntaxError("Unexpected end of input")
        if expected_value is not None and token[0] != expected_value:
            raise SyntaxError(f"Expected '{expected_value}', got '{token[0]}'")
        self.advance()
        return token

    def parse(self) -> Block:
        statements = []
        while self.current_token() is not None:
            stmt = self.parse_statement()
            if stmt:
                statements.append(stmt)
        return Block(statements)

    def parse_statement(self) -> ASTNode | None:
        token = self.current_token()
        if token is None:
            return None

        value, token_type = token

        if value == "if":
            return self.parse_if_statement()
        elif value == "while":
            return self.parse_while_loop()
        elif value == "for":
            return self.parse_for_loop()
        elif value == "return":
            return self.parse_return_statement()
        elif token_type == Token.IDENTIFIER:
            return self.parse_assignment_or_expression()
        else:
            raise SyntaxError(f"Unexpected token: {value}")

    def parse_if_statement(self) -> IfStatement:
        self.consume("if")
        condition = self.parse_expression()
        self.consume("then")

        then_block = []
        while self.current_token() and self.current_token()[0] not in [
            "else",
            "end",
        ]:
            stmt = self.parse_statement()
            if stmt:
                then_block.append(stmt)

        else_block = None
        if self.current_token() and self.current_token()[0] == "else":
            self.consume("else")
            else_block = []
            while self.current_token() and self.current_token()[0] != "end":
                stmt = self.parse_statement()
                if stmt:
                    else_block.append(stmt)

        self.consume("end")
        return IfStatement(
            condition, Block(then_block), Block(else_block) if else_block else None
        )

    def parse_while_loop(self) -> WhileLoop:
        self.consume("while")
        condition = self.parse_expression()
        self.consume("do")

        body = []
        while self.current_token() and self.current_token()[0] != "end":
            stmt = self.parse_statement()
            if stmt:
                body.append(stmt)

        self.consume("end")
        return WhileLoop(condition, Block(body))

    def parse_for_loop(self) -> ForLoop:
        """This should be formatted as for ASSIGN to EXPR do BODY end"""
        self.consume("for")

        assignment: Assignment = self.parse_assignment_or_expression()

        self.consume("to")
        expression = self.parse_expression()
        self.consume("do")

        body = []
        while self.current_token() and self.current_token()[0] != "end":
            stmt = self.parse_statement()
            if stmt:
                body.append(stmt)

        self.consume("end")
        return ForLoop(assignment, expression, Block(body))

    def parse_return_statement(self) -> ReturnStatement:
        self.consume("return")
        value = self.parse_expression()
        return ReturnStatement(value)

    def parse_assignment_or_expression(self) -> ASTNode:
        expr = self.parse_expression()

        if self.current_token() and self.current_token()[0] == "<-":
            self.consume("<-")
            value = self.parse_expression()
            return Assignment(expr, value)

        return expr

    def parse_expression(self) -> ASTNode:
        return self.parse_or_expression()

    def parse_or_expression(self) -> ASTNode:
        left = self.parse_and_expression()

        while self.current_token() and self.current_token()[0] == "or":
            op_token = self.consume()
            right = self.parse_and_expression()
            left = BinaryOp(left, op_token[0], right)

        return left

    def parse_and_expression(self) -> ASTNode:
        left = self.parse_comparison()

        while self.current_token() and self.current_token()[0] == "and":
            op_token = self.consume()
            right = self.parse_comparison()
            left = BinaryOp(left, op_token[0], right)

        return left

    def parse_comparison(self) -> ASTNode:
        left = self.parse_additive()

        while self.current_token() and self.current_token()[0] in [
            "<",
            ">",
            "<=",
            ">=",
            "!=",
            "=",
        ]:
            op_token = self.consume()
            right = self.parse_additive()
            left = BinaryOp(left, op_token[0], right)

        return left

    def parse_additive(self) -> ASTNode:
        left = self.parse_multiplicative()

        while self.current_token() and self.current_token()[0] in ["+", "-"]:
            op_token = self.consume()
            right = self.parse_multiplicative()
            left = BinaryOp(left, op_token[0], right)

        return left

    def parse_multiplicative(self) -> ASTNode:
        left = self.parse_unary()

        while self.current_token() and self.current_token()[0] in ["*", "/"]:
            op_token = self.consume()
            right = self.parse_unary()
            left = BinaryOp(left, op_token[0], right)

        return left

    def parse_unary(self) -> ASTNode:
        if self.current_token() and self.current_token()[0] == "not":
            op_token = self.consume()
            operand = self.parse_unary()
            return UnaryOp(op_token[0], operand)

        return self.parse_postfix()

    def parse_postfix(self) -> ASTNode:
        expr = self.parse_primary()

        while self.current_token():
            token = self.current_token()
            if token[0] == "[":
                self.consume("[")
                index = self.parse_expression()
                self.consume("]")
                expr = ArrayAccess(expr, index)
            else:
                break

        return expr

    def parse_primary(self) -> ASTNode:
        token = self.current_token()
        if token is None:
            raise SyntaxError("Unexpected end of input")

        value, token_type = token

        if token_type == Token.LITERAL:
            self.advance()
            return Literal(value)

        elif token_type == Token.IDENTIFIER:
            self.advance()
            return Identifier(value)

        elif value == "(":
            self.consume("(")
            expr = self.parse_expression()
            self.consume(")")
            return expr

        elif value == "[":
            return self.parse_array_literal()

        else:
            raise SyntaxError(f"Unexpected token: {value}")

    def parse_array_literal(self) -> ArrayLiteral:
        self.consume("[")
        elements = []

        if self.current_token() and self.current_token()[0] != "]":
            elements.append(self.parse_expression())
            while self.current_token() and self.current_token()[0] == ",":
                self.consume(",")
                if self.current_token()[0] != "]":
                    elements.append(self.parse_expression())

        self.consume("]")
        return ArrayLiteral(elements)

File: go/parser.go
package main

import "fmt"

type Node interface{ nodeTag() }

type LiteralNode struct{ Value string }
type IdentNode struct{ Name string }
type ArrayLitNode struct{ Elems []Node }
type ArrayAccessNode struct {
	Array Node
	Index Node
}
type BinaryOpNode struct {
	Left, Right Node
	Op          string
}
type UnaryOpNode struct {
	Op      string
	Operand Node
}
type AssignNode struct {
	Target Node
	Value  Node
}
type IfNode struct {
	Cond Node
	Then *BlockNode
	Else *BlockNode
}
type WhileNode struct {
	Cond Node
	Body *BlockNode
}
type ForNode struct {
	Init      Node
	End       Node
	Body      *BlockNode
	Direction string
	Step      Node
}
type ReturnNode struct{ Value Node }
type BlockNode struct{ Stmts []Node }
type AlgoNode struct {
	Name   string
	Params []string
	Body   *BlockNode
}

func (n *LiteralNode) nodeTag()     {}
func (n *IdentNode) nodeTag()       {}
func (n *ArrayLitNode) nodeTag()    {}
func (n *ArrayAccessNode) nodeTag() {}
func (n *BinaryOpNode) nodeTag()    {}
func (n *UnaryOpNode) nodeTag()     {}
func (n *AssignNode) nodeTag()      {}
func (n *IfNode) nodeTag()          {}
func (n *WhileNode) nodeTag()       {}
func (n *ForNode) nodeTag()         {}
func (n *ReturnNode) nodeTag()      {}
func (n *BlockNode) nodeTag()       {}
func (n *AlgoNode) nodeTag()        {}

type Parser struct {
	tokens []Token
	pos    int
}

func NewParser(tokens []Token) *Parser { return &Parser{tokens: tokens} }

func (p *Parser) cur() *Token {
	if p.pos < len(p.tokens) {
		return &p.tokens[p.pos]
	}
	return nil
}

func (p *Parser) peek(offset int) *Token {
	i := p.pos + offset
	if i < len(p.tokens) {
		return &p.tokens[i]
	}
	return nil
}

func (p *Parser) advance() { p.pos++ }

func (p *Parser) consume(expected string) (Token, error) {
	t := p.cur()
	if t == nil {
		return Token{}, fmt.Errorf("unexpected end of input, expected %q", expected)
	}
	if expected != "" && t.Value != expected {
		return Token{}, fmt.Errorf("expected %q, got %q", expected, t.Value)
	}
	p.advance()
	return *t, nil
}

func (p *Parser) Parse() (*BlockNode, error) {
	var stmts []Node
	for p.cur() != nil {
		s, err := p.parseStatement()
		if err != nil {
			return nil, err
		}
		if s != nil {
			stmts = append(stmts, s)
		}
	}
	return &BlockNode{stmts}, nil
}

func (p *Parser) parseStatement() (Node, error) {
	t := p.cur()
	if t == nil {
		return nil, nil
	}
	switch t.Value {
	case "if":
		return p.parseIf()
	case "while":
		return p.parseWhile()
	case "for":
		return p.parseFor()
	case "return":
		return p.parseReturn()
	case "Algorithm":
		return p.parseAlgo()
	}
	if t.Type == IDENTIFIER {
		return p.parseAssignOrExpr()
	}
	return nil, fmt.Errorf("unexpected token: %q", t.Value)
}

func (p *Parser) parseIf() (Node, error) {
	if _, err := p.consume("if"); err != nil {
		return nil, err
	}
	cond, err := p.parseExpr()
	if err != nil {
		return nil, err
	}
	t := p.cur()
	if t == nil || (t.Value != "then" && t.Value != "do") {
		got := "<EOF>"
		if t != nil {
			got = t.Value
		}
		return nil, fmt.Errorf("expected 'then', got %q", got)
	}
	p.advance()

	var thenStmts []Node
	for p.cur() != nil && p.cur().Value != "end" {
		s, err := p.parseStatement()
		if err != nil {
			return nil, err
		}
		if s != nil {
			thenStmts = append(thenStmts, s)
		}
	}
	if _, err := p.consume("end"); err != nil {
		return nil, err
	}
	return &IfNode{cond, &BlockNode{thenStmts}, nil}, nil
}

func (p *Parser) parseWhile() (Node, error) {
	if _, err := p.consume("while"); err != nil {
		return nil, err
	}
	cond, err := p.parseExpr()
	if err != nil {
		return nil, err
	}
	if _, err := p.consume("do"); err != nil {
		return nil, err
	}
	var stmts []Node
	for p.cur() != nil && p.cur().Value != "end" {
		s, err := p.parseStatement()
		if err != nil {
			return nil, err
		}
		if s != nil {
			stmts = append(stmts, s)
		}
	}
	if _, err := p.consume("end"); err != nil {
		return nil, err
	}
	return &WhileNode{cond, &BlockNode{stmts}}, nil
}

func (p *Parser) parseFor() (Node, error) {
	if _, err := p.consume("for"); err != nil {
		return nil, err
	}
	init, err := p.parseAssignOrExpr()
	if err != nil {
		return nil, err
	}
	if _, err := p.consume("to"); err != nil {
		return nil, err
	}
	end, err := p.parseExpr()
	if err != nil {
		return nil, err
	}
	if _, err := p.consume("do"); err != nil {
		return nil, err
	}
	var stmts []Node
	for p.cur() != nil && p.cur().Value != "end" {
		s, err := p.parseStatement()
		if err != nil {
			return nil, err
		}
		if s != nil {
			stmts = append(stmts, s)
		}
	}
	if _, err := p.consume("end"); err != nil {
		return nil, err
	}
	return &ForNode{init, end, &BlockNode{stmts}, "to", nil}, nil
}

func (p *Parser) parseReturn() (Node, error) {
	if _, err := p.consume("return"); err != nil {
		return nil, err
	}
	val, err := p.parseExpr()
	if err != nil {
		return nil, err
	}
	return &ReturnNode{val}, nil
}

func (p *Parser) parseAlgo() (Node, error) {
	if _, err := p.consume("Algorithm"); err != nil {
		return nil, err
	}
	nameTok, err := p.consume("")
	if err != nil {
		return nil, err
	}
	if nameTok.Type != IDENTIFIER {
		return nil, fmt.Errorf("expected algorithm name, got %q", nameTok.Value)
	}
	if _, err := p.consume("("); err != nil {
		return nil, err
	}
	var params []string
	if p.cur() != nil && p.cur().Value != ")" {
		t, err := p.consume("")
		if err != nil {
			return nil, err
		}
		params = append(params, t.Value)
		for p.cur() != nil && p.cur().Value == "," {
			p.advance()
			t, err = p.consume("")
			if err != nil {
				return nil, err
			}
			params = append(params, t.Value)
		}
	}
	if _, err := p.consume(")"); err != nil {
		return nil, err
	}
	if _, err := p.consume("do"); err != nil {
		return nil, err
	}
	var stmts []Node
	for p.cur() != nil && p.cur().Value != "end" {
		s, err := p.parseStatement()
		if err != nil {
			return nil, err
		}
		if s != nil {
			stmts = append(stmts, s)
		}
	}
	if _, err := p.consume("end"); err != nil {
		return nil, err
	}
	return &AlgoNode{nameTok.Value, params, &BlockNode{stmts}}, nil
}

func (p *Parser) parseAssignOrExpr() (Node, error) {
	expr, err := p.parseExpr()
	if err != nil {
		return nil, err
	}
	if p.cur() != nil && p.cur().Value == "<-" {
		p.advance()
		val, err := p.parseExpr()
		if err != nil {
			return nil, err
		}
		return &AssignNode{expr, val}, nil
	}
	return expr, nil
}

func (p *Parser) parseExpr() (Node, error) { return p.parseAdd() }

func (p *Parser) parseAdd() (Node, error) {
	left, err := p.parseMul()
	if err != nil {
		return nil, err
	}
	for p.cur() != nil && (p.cur().Value == "+" || p.cur().Value == "-") {
		op := p.cur().Value
		p.advance()
		right, err := p.parseMul()
		if err != nil {
			return nil, err
		}
		left = &BinaryOpNode{left, right, op}
	}
	return left, nil
}

func (p *Parser) parseMul() (Node, error) {
	left, err := p.parseUnary()
	if err != nil {
		return nil, err
	}
	for p.cur() != nil && (p.cur().Value == "*" || p.cur().Value == "/") {
		op := p.cur().Value
		p.advance()
		right, err := p.parseUnary()
		if err != nil {
			return nil, err
		}
		left = &BinaryOpNode{left, right, op}
	}
	return left, nil
}

func (p *Parser) parseUnary() (Node, error) {
	return p.parsePostfix()
}

func (p *Parser) parsePostfix() (Node, error) {
	expr, err := p.parsePrimary()
	if err != nil {
		return nil, err
	}
	for p.cur() != nil && p.cur().Value == "[" {
		p.advance()
		idx, err := p.parseExpr()
		if err != nil {
			return nil, err
		}
		if _, err := p.consume("]"); err != nil {
			return nil, err
		}
		expr = &ArrayAccessNode{expr, idx}
	}
	return expr, nil
}

func (p *Parser) parsePrimary() (Node, error) {
	t := p.cur()
	if t == nil {
		return nil, fmt.Errorf("unexpected end of input")
	}
	if t.Type == LITERAL {
		p.advance()
		return &LiteralNode{t.Value}, nil
	}
	if t.Type == IDENTIFIER {
		p.advance()
		return &IdentNode{t.Value}, nil
	}
	if t.Value == "(" {
		p.advance()
		expr, err := p.parseExpr()
		if err != nil {
			return nil, err
		}
		if _, err := p.consume(")"); err != nil {
			return nil, err
		}
		return expr, nil
	}
	if t.Value == "[" {
		return p.parseArrayLit()
	}
	return nil, fmt.Errorf("unexpected token in expression: %q", t.Value)
}

func (p *Parser) parseArrayLit() (Node, error) {
	p.advance()
	var elems []Node
	if p.cur() != nil && p.cur().Value != "]" {
		e, err := p.parseExpr()
		if err != nil {
			return nil, err
		}
		elems = append(elems, e)
		for p.cur() != nil && p.cur().Value == "," {
			p.advance()
			if p.cur() != nil && p.cur().Value == "]" {
				break
			}
			e, err = p.parseExpr()
			if err != nil {
				return nil, err
			}
			elems = append(elems, e)
		}
	}
	if _, err := p.consume("]"); err != nil {
		return nil, err
	}
	return &ArrayLitNode{elems}, nil
}
